MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c

```diff
@@ -0,0 +1,473 @@
+          //
+          NewFlbas                = NewNamespaceData->Flbas;
+          LbaFmtIdx               = NewFlbas & NVME_LBA_FORMATNVM_LBAF_MASK;
+          Lbads                   = NewNamespaceData->LbaFormat[LbaFmtIdx].Lbads;
+          Device->Media.BlockSize = (UINT32)1 << Lbads;
+          Device->Media.LastBlock = NewNamespaceData->Nsze - 1;
+
+          CopyMem (&Device->NamespaceData, NewNamespaceData, sizeof (NVME_ADMIN_NAMESPACE_DATA));
+        }
+      }
+    }
+  }
+
+  return Status;
+}
+
+/**
+  Send NVM Express Sanitize Admin Command
+
+  The Sanitize command is used to start a sanitize operation or to recover from a previously
+  failed sanitize operation. The sanitize operation types that may be supported are Block
+  Erase, Crypto Erase, and Overwrite.
+
+  All sanitize operations are processed in the background (i.e., completion of the Sanitize
+  command does not indicate completion of the sanitize operation).
+
+  @param[in] This                Indicates a pointer to the calling context (Block IO Protocol)
+  @param[in] NamespaceId         The NVM Express namespace ID  for which a device path node is to be
+                                 allocated and built. Caller must set the NamespaceId to zero if the
+                                 device path node will contain a valid UUID.
+  @param[in] SanitizeAction      Sanitize action
+  @param[in] NoDeallocAfterSani  No deallocate after sanitize option
+  @param[in] OverwritePatter     Pattern to overwrite old user data
+
+  @retval EFI_SUCCESS            All the namespaces in the device are successfully enumerated.
+  @return Others                 Some error occurs when enumerating the namespaces.
+
+ **/
+EFI_STATUS
+NvmExpressSanitize (
+  IN EFI_BLOCK_IO_PROTOCOL  *This,
+  IN UINT32                 NamespaceId,
+  IN UINT32                 SanitizeAction,
+  IN UINT32                 NoDeallocAfterSanitize,
+  IN UINT32                 OverwritePattern
+  )
+{
+  NVME_DEVICE_PRIVATE_DATA                  *Device;
+  EFI_NVM_EXPRESS_PASS_THRU_COMMAND_PACKET  CommandPacket;
+  EFI_NVM_EXPRESS_COMMAND                   Command;
+  EFI_NVM_EXPRESS_COMPLETION                Completion;
+  NVME_ADMIN_SANITIZE                       SanitizeCdw10Cdw11;
+  EFI_STATUS                                Status;
+
+  Device = NVME_DEVICE_PRIVATE_DATA_FROM_BLOCK_IO (This);
+
+  ZeroMem (&CommandPacket, sizeof (EFI_NVM_EXPRESS_PASS_THRU_COMMAND_PACKET));
+  ZeroMem (&Command, sizeof (EFI_NVM_EXPRESS_COMMAND));
+  ZeroMem (&Completion, sizeof (EFI_NVM_EXPRESS_COMPLETION));
+  ZeroMem (&SanitizeCdw10Cdw11, sizeof (NVME_ADMIN_SANITIZE));
+
+  CommandPacket.NvmeCmd        = &Command;
+  CommandPacket.NvmeCompletion = &Completion;
+  CommandPacket.CommandTimeout = NVME_GENERIC_TIMEOUT;
+  CommandPacket.QueueType      = NVME_ADMIN_QUEUE;
+  Command.Cdw0.Opcode          = NVME_ADMIN_SANITIZE_CMD;
+  Command.Nsid                 = NamespaceId;
+
+  DEBUG ((
+    DEBUG_ERROR,

```

Comment:
Is this really an error?

Reply:
No, definitely not an error. I removed a check, but kept the debug statement. I will update to DEBUG_VERBOSE or DEBUG_INFO.